이미지 파일 업로드, 다운로드 기능 구현

ghk71 opened the pull request:
-post로 파일을 업로드 가능
-get으로 업로드 된 파일 다운가능

**파일 업로드 방법
Postman이용 -> Post로 바꾸고 http://localhost:8080/uploadFile 그 후 Body -> form-data -> 체크박스 체크, KEY에 file 입력, VALUE로 파일 선택 -> 파일 업로드 완료 -> Uploads폴더 자동 생성

**파일 받는 방법
Postman이용 -> Get으로 바꾸고 http://localhost:8080/downloadFile/파일이름

ghk71 committed:
이미지 파일 업로드, 다운로드 기능 구현
-post로 파일을 업로드 가능
-get으로 업로드 된 파일 다운가능

**파일 업로드 방법
Postman이용 -> Post로 바꾸고 http://localhost:8080/uploadFile 그 후 Body -> form-data -> 체크박스 체크, KEY에 file 입력, VALUE로 파일 선택 -> 파일 업로드 완료 -> Uploads폴더 자동 생성

**파일 받는 방법
Postman이용 -> Get으로 바꾸고 http://localhost:8080/downloadFile/파일이름

parkjonghyeob commented on coin-server/src/main/java/kr/or/teamserver/coinserver/controller/FileController.java:
```diff
@@ -0,0 +1,60 @@
+import javax.servlet.http.HttpServletRequest;
+import java.io.IOException;
+
+@RestController
```
`@RequestMapping("/file")` 추가

parkjonghyeob commented on coin-server/src/main/java/kr/or/teamserver/coinserver/controller/FileController.java:
```diff
@@ -0,0 +1,60 @@
+
+    @PostMapping("/uploadFile")
+    @ResponseBody
+    public UploadFileResponse uploadFile(@RequestParam("file") MultipartFile file) {
```
`@PostMapping("/upload")`
`uploadFile` -> `upload` (이미 파일 클래스니까 굳이 suffix 는 필요 없을듯

parkjonghyeob replied:
method 명이 변경이 안되었네.

parkjonghyeob commented on coin-server/src/main/java/kr/or/teamserver/coinserver/controller/FileController.java:
```diff
@@ -0,0 +1,60 @@
+        String fileDownloadUri = ServletUriComponentsBuilder.fromCurrentContextPath()
+                .path("/downloadFile/")
+                .path(fileName)
+                .toUriString();
```
```
String fileName = fileStorageService.storeFile(file);
...
               .toUriString();
```
FileService 를 만들어서 거기서 구현

parkjonghyeob commented on coin-server/src/main/java/kr/or/teamserver/coinserver/controller/FileController.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @GetMapping("/downloadFile/{fileName:.+}")
+    public ResponseEntity<Resource> downloadFile(@PathVariable String fileName, HttpServletRequest request) {
```
위에 업로드와 동일한 부분 수정

parkjonghyeob commented on coin-server/src/main/java/kr/or/teamserver/coinserver/controller/FileController.java:
```diff
@@ -0,0 +1,60 @@
+
+        if(contentType == null) {
+            contentType = "application/octet-stream";
+        }
```
```
String contentType = null;
...
if(contentType == null) {
    contentType = "application/octet-stream";
}
```
FileService 에서 처리. 그리고 null 을 초기값으로 넣지 말고, default 를 `application/octet-stream` 으로

parkjonghyeob commented on coin-server/src/main/java/kr/or/teamserver/coinserver/payload/UploadFileResponse.java:
```diff
@@ -0,0 +1,17 @@
+package kr.or.teamserver.coinserver.payload;
+
+public class UploadFileResponse {
+    private String fileName;
```
private final

parkjonghyeob commented on coin-server/src/main/java/kr/or/teamserver/coinserver/property/FileStorageProperties.java:
```diff
@@ -0,0 +1,16 @@
+
+@ConfigurationProperties(prefix = "file")
+public class FileStorageProperties {
+    private String uploadDir;
```
baseDir 은 어떨까?

parkjonghyeob commented on coin-server/src/main/java/kr/or/teamserver/coinserver/service/FileStorageService.java:
```diff
@@ -0,0 +1,67 @@
+
+        try {
+            if(fileName.contains("..")) {
+                throw new FileStorageException("Sorry! Filename contains invalid path sequence " + fileName);
```
IllegalArgumentException 같은 것도 괜찮겠네..

parkjonghyeob commented on coin-server/src/main/java/kr/or/teamserver/coinserver/service/FileStorageService.java:
```diff
@@ -0,0 +1,67 @@
+            } else {
+                throw new MyFileNotFoundException("File not found " + fileName);
+            }
+        } catch (MalformedURLException ex) {
```
여기를 try-catch 한 이유가 있을까?

parkjonghyeob commented on coin-server/src/main/resources/application.properties:
```diff
@@ -2,4 +2,19 @@ spring.datasource.url=jdbc:h2:mem:washer
+
+
+## MULTIPART (MultipartProperties)
+# Enable multipart uploads
```
나중에 주석들은 삭제

ghk71 committed:
@RequestMapping 추가
@GetMapping의 file주소 삭제

final 변수 선언 추가

application.properties의 주석 삭제

uploadDir -> baseDir

ghk71 committed:
fileService 추가

ghk71 committed:
FileService 업로드, 다운로드 문제 해결
-FileStorageService -> File Service
-기존의 FileService의 내용 옮긴 후 삭제

ghk71 committed:
ExceptionHandlingController 생성
exception 처리

parkjonghyeob commented on coin-server/src/main/java/kr/or/teamserver/coinserver/payload/UploadFileResponse.java:
```diff
@@ -1,10 +1,10 @@
-    private String fileType;
-    private long size;
+    private final String fileName;
+    private final String fileDownloadUri;
```
fileDownloadUrl 이 아닐까?

parkjonghyeob commented on coin-server/src/main/java/kr/or/teamserver/coinserver/service/FileService.java:
```diff
@@ -0,0 +1,45 @@
+        return fileDownloadUri;
+    }
+
+    public String getContentType(Resource resource, HttpServletRequest request) {
```
왠만하면 서비스쪽으로 Request 나 Response 를 가지고 오지 않는게 좋을 듯~

parkjonghyeob commented on coin-server/src/main/java/kr/or/teamserver/coinserver/service/FileService.java:
```diff
@@ -84,8 +84,8 @@ public String getContentType(Resource resource, HttpServletRequest request) {
-        } catch (IOException ex) {
-            logger.info("Could not determine file type.");
+        } catch (Exception ex) {
+            throw new MyFileNotFoundException("Could not determine file type.", ex);
```
그냥 FileNotFoundException 이면 될듯..

ghk71 committed:
uri -> url
FileNotFoundException으로 변경

ghk71 committed:
Exception 오류 수정

parkjonghyeob commented on coin-server/src/main/java/kr/or/teamserver/coinserver/controller/FileController.java:
```diff
@@ -0,0 +1,47 @@
+    @PostMapping("/upload")
+    @ResponseBody
+    public UploadFileResponse uploadFile(@RequestParam("file") MultipartFile file) {
+        String fileName = fileService.storeFile(file);
```
FileName 과 FileDownloadUrl 을 모두 가지는 객체를 리턴하는건 어떨까?

parkjonghyeob commented on coin-server/src/main/java/kr/or/teamserver/coinserver/exception/FileNotFoundException.java:
```diff
@@ -0,0 +1,15 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.NOT_FOUND)
```
이건 여기 없어도 괜찮아.
ExceptionHandler 쪽에 있으면 됨

parkjonghyeob commented on coin-server/src/main/java/kr/or/teamserver/coinserver/payload/UploadFileResponse.java:
```diff
@@ -0,0 +1,31 @@
+
+    public long getSize() {
+        return size;
+    }
```
toString() 은 오버라이드 해주면 좋을 듯~~

ghk71 committed:
Exception 수정 gettersetter 생성

ghk71 committed:
Merge branch 'develop' into feature/Image
